subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/verification/DependencyVerificationIntegrityCheckIntegTest.groovy

```diff
@@ -0,0 +1,384 @@
+
+        then:
+        outputContains("Dependency verification is an incubating feature.")
+
+        when:
+        fails ":compileJava"
+
+        then:
+        failure.assertHasCause("""Dependency verification failed for configuration ':compileClasspath':
+  - On artifact foo-1.0.jar (org:foo:1.0): expected a '$kind' checksum of 'invalid' but was '$value'
+This can indicate that a dependency has been compromised. Please verify carefully the checksums.""")
+
+        where:
+        kind     | value
+        "md5"    | "ea8b622874eaa501476e0ebbe0c562ed"
+        "sha1"   | "16e066e005a935ac60f06216115436ab97c5da02"
+        "sha256" | "20ae575ede776e5e06ee6b168652d11ee23069e92de110fdec13fbeaa5cf3bbc"
+        "sha512" | "734fce768f0e1a3aec423cb4804e5cdf343fd317418a5da1adc825256805c5cad9026a3e927ae43ecc12d378ce8f45cc3e16ade9114c9a147fda3958d357a85b"
+    }
+
+    def "can collect multiple errors in a single dependency graph"() {
+        createMetadataFile {
+            addChecksum("org:foo:1.0", "sha1", "invalid")
+            addChecksum("org:bar:1.0", "sha1", "also invalid")
+            addChecksum("org:baz:1.0", "sha1", "c554a4a45e3ed3da494befb446fb2923b8bcecef")
+        }
+
+        given:
+        javaLibrary()
+        uncheckedModule("org", "foo", "1.0") {
+            dependsOn("org", "bar", "1.0")
+            dependsOn("org", "baz", "1.0")
+        }
+        uncheckedModule("org", "bar")
+        uncheckedModule("org", "baz")
+        buildFile << """
+            dependencies {
+                implementation "org:foo:1.0"
+            }
+        """
+
+        when:
+        fails ":compileJava"
+
+        then:
+        failure.assertHasCause("""Dependency verification failed for configuration ':compileClasspath':
+  - On artifact bar-1.0.jar (org:bar:1.0): expected a 'sha1' checksum of 'also invalid' but was '42077067b52edb41c658839ab62a616740417814'
+  - On artifact foo-1.0.jar (org:foo:1.0): expected a 'sha1' checksum of 'invalid' but was '16e066e005a935ac60f06216115436ab97c5da02'
+This can indicate that a dependency has been compromised. Please verify carefully the checksums.""")
+    }
+
+    @ToBeFixedForInstantExecution
+    def "fails on the first access to an artifact (not at the end of the build)"() {
+        createMetadataFile {
+            addChecksum("org:foo:1.0", "sha1", "invalid")
+        }
+
+        given:
+        javaLibrary()
+        uncheckedModule("org", "foo")
+        uncheckedModule("org", "bar")
+        buildFile << """
+            dependencies {
+                implementation "org:foo:1.0"
+                testImplementation "org:bar:1.0"
+            }
+
+            task resolve {
+                inputs.files(configurations.compileClasspath)
+                inputs.files(configurations.testRuntimeClasspath)
+                doLast {
+                    println "First resolution"
+                    println configurations.compileClasspath.files
```

Comment:
❓ What happens if e.g. some build logic uses `configurations.compileClasspath.iterator().next()` consuming only the first artifact for which the checksum is absent/wrong?

Reply:
AFAIK there's no partial resolution so it will fail with the same error.

Reply:
Good 👍
What about capturing that in the test? I think coverage should be very defensive here.

Reply:
Added more coverage

Reply:
Much better 👍 I trust your judgment about the set of ways this should be triggered. The test setup/assertion is a bit convoluted, no strong opinion but it may be made more readable.